lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.cpp

```diff
@@ -2225,7 +2225,8 @@ void ObjectFileMachO::ParseSymtab(Symtab &symtab) {
   const char *file_name = file.GetFilename().AsCString("<Unknown>");
   LLDB_SCOPED_TIMERF("ObjectFileMachO::ParseSymtab () module = %s", file_name);
   LLDB_LOG(log, "Parsing symbol table for {0}", file_name);
-  Progress progress(llvm::formatv("Parsing symbol table for {0}", file_name));
+  Progress progress(llvm::formatv("Parsing symbol table for {0}", file_name),
+                    Progress::ProgressReportType::eAggregateProgressReport);
```

Comment:
We could add a progress category as a string instead of a report type? Then this would be easy to group all of symbol table parsing into one section. 
```
std::string category("Symbol table parsing");
Progress progress(llvm::formatv("Parsing symbol table for {0}", file_name), category);
```
I used a local variable for the category for clarity

Reply:
When the functionality to add details to a progress report [was added](https://reviews.llvm.org/D143690) in LLDB we wanted to group operations under large progress reports and then incrementally update them but we found that there was no easy place to do that for symbol table and debug info operations so for now, we want to us a flag to indicate to a UI how these operations are reported. 

We did find that it was possible to group Swift-based operations into their own large progress reports (https://github.com/apple/llvm-project/pull/7769) and so we'd want to have these operations be the "non-aggregate/umbrella" operations.

Reply:
Adding a string to represent the category (one of a possible vector of strings for each operation even?) looks like a good idea though.

Reply:
> Adding a string to represent the category (one of a possible vector of strings for each operation even?) looks like a good idea though.

Seems to me that a `category` string that could be anything ("Parsing symbol tables" or "Indexing DWARF" or "Parsing DWARF") would allow us to group similar notifications together in an IDE better than a single `ProgressReportType::eAggregateProgressReport` vs `ProgressReportType::eNonAggregateProgressReport` enum value. With this solution, does this mean that all notifications that with `is_aggregate == true` would show up in one area and `is_aggregate == false` would show up another area? If we use a `category` string, any concurrent notifications that share the same `category` could show up under some UI that displayed the `category` string and allowed it to be expanded

Reply:
The intention is that if `is_aggregate == false` then the progress report type can be grouped together in LLDB under one large umbrella progress report for that operation whereas if `is_aggregate == true` then the IDE could potentially filter out those progress reports since each progress report happening individually means that they look spammy in an IDE.

Instead of using a string as a `category` field we could have each progress report type/category be an enum in `lldb_enumerations.h` i.e. `ProgressReportType/Category == eProgressSymbolTableParse` or `eProgressDWARFIndexing`

Reply:
@clayborg We're also under a little time pressure on this patch so I'd be happy to fix up all major issues and discuss further in order to get this in :)

Reply:
I totally understand about timeframes and getting this in ASAP.

I am just trying to understand what the goal is here as what this change essentially does is groups all the progress notifications for all progress reports into the `is_aggregate == true` except for the SymbolLocationDefault's progress for downloading dSYM files. So essentially all progress notifications are considered spammy and something that won't get displayed? An enum means anytime someone adds a new progress, they need to make a category enum for it. I still think either a category string still makes the most sense. Then an IDE can watch for progress notifications, and if many come in from the same category, it can show a spinning progress with the category name "Parsing symbol tables" and any incoming progress updates could keep this progress UI alive. If a user opens up this top level "Parsing symbol tables" category, the individual progresses that are alive can be shown in detail. This allows an IDE to watch the progress notifications and if it gets too many too quickly from a specific category, it can group them all under the category name and not have to show specific individual entries that come and go too quickly.

The code from `SymbolFileDWARF::InitializeObject()` is what I am guessing can get really spammy for both the apple tables and DWARF5 tables. There is no work to be done when loading these tables, so I am guessing the progress is really quick for those and not that useful. We could remove these if they are some of the offending progress notifications that are causing issues.

Manual DWARF parsing and symbol table parsing, those do take some time, so they tend to be very useful as they really slow down debugger startup when you have large files and when users are sitting there watching a blank IDE window and no visible progress is made by the debugger, it can lead them to kill the debug session if there is no actual feedback.

Reply:
If the goal is to get rid of super spammy notifications (ones that are too quick), that can actually be done client side when the progress notification events are received by setting a timeout once the start progress has been received, and if a certain timeout happens (more than 1 second), then just don't show the progress, but if it does last then pop the progress dialog up. We tried this in lldb-dap.cpp. See the function named `void ProgressEventThreadFunction()` which is:

```
void DAP::SendProgressEvent(uint64_t progress_id, const char *message,
                            uint64_t completed, uint64_t total) {
  progress_event_reporter.Push(progress_id, message, completed, total);
}
```
We have a class called `ProgressEventReporter`:
```
/// Class that filters out progress event messages that shouldn't be reported
/// to the IDE, because they are invalid, they carry no new information, or they
/// don't last long enough.
///
/// We need to limit the amount of events that are sent to the IDE, as they slow
/// the render thread of the UI user, and they end up spamming the DAP
/// connection, which also takes some processing time out of the IDE.
class ProgressEventReporter {
```
The problem we ran into with sending tons of notifications through the debug adaptor protocol was it was too many packets, so we trimmed them down. The main issue with this approach is if you get 1000 progress notifications and each progress takes less than the minimum report time, like 1 second, and if each progress takes just under the timeout of 1 second, you have a long delay with no feedback.

For progress notifications like the DWARF manual index, it has a total count of the number of compile units and each time a compile unit is indexed, the progress gets advanced and can cause a ton of update packets to be sent, this class helps reduce these. 

So the `ProgressEventReporter` class will:
- not report events if they are too quick, but if they exceed a minimum threshold, they get report, albeit a bit late
- if there is a progress that is really spammy (1 of 10000, 2 of 10000, 3 of 10000, etc) it will report progress only every so often. This really reduces the number of progress notifications.

Reply:
When we looked at how to group progress notifications by `is_aggregate` we found that although the symbol table/DWARF progress notifications would be grouped under `is_aggregate == true`, long-running operations related to Swift (importing/loading Swift modules etc.) could actually be grouped under `is_aggregate == false` which allows the IDE team to display and group those notifications. (https://github.com/apple/llvm-project/pull/7769). We want to integrate this enum flag into the Swift progress reports once landed. 

That being said, this does group long-running LLDB notifications as all being too spammy to show since we couldn't group them together under umbrella progress reports. We don't want to keep it this way forever as ultimately we'd like to group all categories of progress reports under their own umbrella `Progress` objects instead of sending several individual progress reports, but this is what we're doing right now to help the IDE have more streamlined UI progress reports.

We also thought about going with the timeout approach that you suggested but ran into similar issues as you mentioned wherein if many reports were too short then they wouldn't get displayed.

Reply:
@clayborg I understand your point but how would a category string be different from filtering out the progress events based on the progress title string (or parts of it) ?

I see how certain progress reports could be relevant to linux where we wouldn't want to surface it on macOS. The reason we came up with this approach is precisely because we prefer having LLDB tell its clients what progress should be displayed, rather than having the filtering logic be done in the client.

May we could keep a list of blacklisted progress title / categories for every platform and either submit progress reports if they're not in that list or have an SBAPI to fetch the list of blacklisted progresses from the client ? What do you think about this ?

(also cc @JDevlieghere since he will probably have some opinions about this).

Reply:
> @clayborg I understand your point but how would a category string be different from filtering out the progress events based on the progress title string (or parts of it) ?

It would stop clients from having to add logic to try and look for common substrings, but other than that not much different. It would allow progress titles to vary quite a bit more, but they don't right now.

> I see how certain progress reports could be relevant to linux where we wouldn't want to surface it on macOS. The reason we came up with this approach is precisely because we prefer having LLDB tell its clients what progress should be displayed, rather than having the filtering logic be done in the client.

Not sure I follow? Are you saying if `is_aggregate == true` that we are letting users know they should ignore these progress reports? And we are setting to true for all progress reports except for Swift and a few other rarely used ones?

> May we could keep a list of blacklisted progress title / categories for every platform and either submit progress reports if they're not in that list or have an SBAPI to fetch the list of blacklisted progresses from the client ? What do you think about this ?

One ideas is if we have categories, we could have settings that could control if the progress reports for a given category should be displayed. Those could easily be set after initializing a debugger object.

> (also cc @JDevlieghere since he will probably have some opinions about this).

Sounds good

The entire idea of the progress dialogs is to provide insight for when the debugger seems to be stalled, but it is actually doing work. If we have any progress items that are not useful, we should remove them. But parsing symbol tables, even on Mac, can sometimes take a few seconds on larger binaries, so I find this useful when running LLDB from the command line. Indexing DWARF is a huge time sync for really large projects which often can take many minutes to complete and when I load them on the command line, seeing the progress in the terminal is very useful.

So a few things that seem to stand out to me:
- Darwin will never see the `Manually indexing DWARF for <path>` progress dialogs, and these are needed and should always be displayed.
- Darwin does see the `Parsing symbol table for <path> progress dialogs, but many are fast and probably are part of the problem you want to fix?
- "Loading Apple DWARF index for <path>" has no real work being done here, just initializing a memory buffer, seems like it should probably be removed to cut down on the noise?
- "Loading DWARF5 index for <path>" is the same kind of thing for both Darwin and linux (no real work to be done, the accelerator tables are "ready to use")
- "Locating external symbol file for <path>" seems useful if the download can take a while

I don't see any other uses in the upstream LLDB sources.

But my main point is we make Progress dialogs for a reason and we should make sure they are useful and that everyone wants them.

Reply:
And if we want to allow different platforms to say what is useful and what isn't we can provide commands to control these if needed. We can have an internal enumeration for the categories if needed, or we can add a string as the category to help say "please disable all symbol table parsing" progress dialogs.